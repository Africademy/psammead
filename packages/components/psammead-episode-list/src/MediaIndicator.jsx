import React from 'react';
import styled, { css } from 'styled-components';
import { string } from 'prop-types';

const Wrapper = styled.div`
  display: inline-block;
  ${({ size }) => css`
    width: ${size};
    height: ${size};
  `}
`;

const MediaIndicator = ({ size }) => (
  <Wrapper
    size={size}
    dangerouslySetInnerHTML={{
      __html: `
      <svg class="rounded-play-button" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 40 40">
          <path class="rounded-play-button__outer-circle" d="M20,0C8.954,0,0,8.954,0,20s8.954,20,20,20s20-8.954,20-20S31.046,0,20,0z"/>
          <path class="rounded-play-button__inner-circle" fill="#fff" d="M20,1.765C9.929,1.765,1.765,9.929,1.765,20S9.929,38.235,20,38.235S38.235,30.071,38.235,20
          S30.071,1.765,20,1.765z"/>
          <polygon class="rounded-play-button__triangle" points="15.799,26.94 27.062,20.009 15.799,13.078 "/>
      </svg>
      `,
    }}
  />
);

MediaIndicator.propTypes = {
  size: string.isRequired,
};

export default MediaIndicator;
