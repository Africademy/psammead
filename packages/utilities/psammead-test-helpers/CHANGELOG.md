# Psammead Test Helpers Changelog

| Version | Description |
|---------|-------------|
| 0.1.2   | [PR#255](https://github.com/BBC-News/psammead/pull/255) Bumps package version to fix what appears to be an erroneously-published version. |
| 0.1.2   | [PR#212](https://github.com/BBC-News/psammead/pull/212) Update package description and README. |
| 0.1.1   | [PR#173](https://github.com/BBC-News/psammead/pull/173) Update PRs welcome link |
| 0.1.0   | [PR#80](https://github.com/BBC-News/psammead/pull/80) Create initial package, pulled in from [simorgh](https://github.com/BBC-News/simorgh). |
