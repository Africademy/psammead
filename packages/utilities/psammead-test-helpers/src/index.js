import { render, waitForDomChange } from '@testing-library/react';
import 'jest-styled-components';
import deepClone from 'ramda/src/clone';

const renderWithHelmet = async component => {
  /*
 * Similar to this problem https://github.com/testing-library/react-testing-library/issues/402
 * This helper was created to solve the problem of rendering helmet/head contents in snapshots.
 * Pass in a component that uses helmet somewhere in the component tree.
 * The full html tree is rendered and returned which you can then use to snapshot helmet/head contents.
 * Example of use in a test file:
 it('should render correctly', async () => {
    const html = await renderHelmet(<SomeComponent {...someProps} />);
    expect(html).toMatchSnapshot();
  });
 */

  const { container } = render(component);
  const noop = () => {};
  const ARBITRARY_TIMEOUT = 100; // seems long enough for any dom mutations to occur

  return waitForDomChange({
    container: document.querySelector('head'),
    timeout: ARBITRARY_TIMEOUT,
  })
    .catch(noop) // handle a waitForDomChange timeout
    .then(mutationsList => {
      const domMutationDetected = mutationsList && mutationsList.length;

      if (domMutationDetected) {
        // helmet was probably used so we should get the full html

        const htmlElement = document.querySelector('html');

        const helmetElements = document.querySelectorAll(
          '[data-react-helmet="true"]',
        );

        const removeHelmetAttributes = el =>
          el.removeAttribute('data-react-helmet'); // remove react-helmet attribute noise from elements

        removeHelmetAttributes(htmlElement); // remove react-helmet attribute noise from elements

        Array.from(helmetElements).forEach(removeHelmetAttributes);

        return { container: document.querySelector('html') };
      }

      return { container };
    });
};

export const shouldMatchSnapshot = (title, component) => {
  it(title, async () => {
    // select the first child to remove the pointless wrapping div from snapshots
    const removeWrappingDiv = container => container.firstChild;
    const { container } = await renderWithHelmet(component);
    const hasOneChild = container.children.length === 1;
    /*
     * if the container has more than one child then it's a component that uses a
     * fragment at the top level so we should not select the first child because it
     * wouldn't snapshot the whole component
     */
    expect(
      hasOneChild ? removeWrappingDiv(container) : container,
    ).toMatchSnapshot();
  });
};

export const isNull = (title, component) => {
  it(title, () => {
    const { container } = render(component);
    expect(container.firstChild).toBeNull();
  });
};

export const setWindowValue = (key, value) => {
  const windowValue = window[key];
  delete window[key];

  let newValue = value;

  if (value && typeof value === 'object') {
    newValue = {
      ...deepClone(windowValue),
      ...value,
    };
  }

  Object.defineProperty(window, key, {
    value: newValue,
    writable: true,
  });
};

export const resetWindowValue = (key, value) => {
  Object.defineProperty(window, key, {
    value,
    writable: true,
  });
};

export const suppressPropWarnings = warnings => {
  const { expectedWarnings } = window;
  if (expectedWarnings && Array.isArray(expectedWarnings)) {
    window.expectedWarnings = [...expectedWarnings, warnings];
  } else {
    window.expectedWarnings = [warnings];
  }
};

const errorIfMissingKey = (keys, object, message) => {
  keys.forEach(key => {
    if (!(key in object)) {
      throw new Error(`Missing value '${key}' in ${message}.`);
    }
  });
};

const checkKeysExistInBothObjects = (object1, object2, message1, message2) => {
  const object1Keys = Object.keys(object1);
  const object2Keys = Object.keys(object2);

  errorIfMissingKey(object1Keys, object2, message2);
  errorIfMissingKey(object2Keys, object1, message1);
};

const checkTypesOfExports = (
  actualExportsByName,
  actualExports,
  expectedExports,
  utilityName,
) => {
  actualExportsByName.forEach(actualExportName => {
    const actualExportValue = actualExports[utilityName][actualExportName];
    const expectedExport = expectedExports[utilityName][actualExportName];
    const typeCheck = typeof actualExportValue === expectedExport; // eslint-disable-line valid-typeof

    // if this fails it is likely that an export is missing from the unit test expectation
    expect(typeCheck).toBe(true);
  });
};

export const testUtilityPackages = (
  actualExports,
  expectedExports,
  packageName,
) => {
  const actualUtilities = Object.keys(actualExports);

  // check if the actual exported file has defined expected values to match and that the expected values are actually exported
  checkKeysExistInBothObjects(
    actualExports,
    expectedExports,
    `the actual utilities for '${packageName}'`,
    `the expected utilities for '${packageName}'`,
  );

  actualUtilities.forEach(utilityName => {
    const actualExportsByName = Object.keys(actualExports[utilityName]);

    // check if each of the actual exported consts have an expected value to match and that the expected values are actually exported
    checkKeysExistInBothObjects(
      actualExports[utilityName],
      expectedExports[utilityName],
      `the actual export for '${packageName}/${utilityName}'`,
      `the expected export for '${packageName}/${utilityName}'`,
    );

    checkTypesOfExports(
      actualExportsByName,
      actualExports,
      expectedExports,
      utilityName,
    );
  });
};
