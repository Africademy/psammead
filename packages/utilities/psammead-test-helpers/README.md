# psammead-test-helpers &middot; [![GitHub license](https://img.shields.io/badge/license-Apache%202.0-blue.svg)](https://github.com/BBC-News/psammead/blob/latest/LICENSE) [![npm version](https://img.shields.io/npm/v/@bbc/psammead-test-helpers.svg)](https://www.npmjs.com/package/@bbc/psammead-test-helpers) [![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg)](https://reactjs.org/docs/how-to-contribute.html#your-first-pull-request)

This package is a collection helper methods used for snapshot testing, with support for styled components.

## Installation

```jsx
npm install @bbc/psammead-test-helpers --save
```

## Usage

```jsx
import { shouldMatchSnapshot } from '@bbc/psammead-test-helpers';

shouldMatchSnapshot(
  'should render correctly',
  <h1>Hello World</h1>,
);
```

## Roadmap

There is currently a plan to migrate from `react-test-renderer` to `enzyme` as it provides a nicer higher-level API over the same functionality, as well as having support for more granular unit tests.

## Contributing

Psammead is completely open source. We are grateful for any contributions, whether they be new components, bug fixes or general improvements. Please see our primary contributing guide which can be found at [the root of the Psammead respository](https://github.com/BBC-News/psammead/blob/latest/CONTRIBUTING.md).

### [Code of Conduct](https://github.com/BBC-News/psammead/blob/latest/CODE_OF_CONDUCT.md)

We welcome feedback and help on this work. By participating in this project, you agree to abide by the [code of conduct](https://github.com/BBC-News/psammead/blob/latest/CODE_OF_CONDUCT.md). Please take a moment to read it.

### License

Psammead is [Apache 2.0 licensed](https://github.com/BBC-News/psammead/blob/latest/LICENSE).
