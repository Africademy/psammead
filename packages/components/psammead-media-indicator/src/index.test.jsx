import React from 'react';
import { shouldMatchSnapshot } from '@bbc/psammead-test-helpers';
import MediaIndicator from './index';

describe('MediaIndicator', () => {
  shouldMatchSnapshot(
    'should render correctly',
    <MediaIndicator duration="2:15" />,
  );
});
