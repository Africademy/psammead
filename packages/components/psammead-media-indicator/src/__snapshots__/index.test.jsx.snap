// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MediaIndicator should render correctly 1`] = `
.c3 {
  -webkit-clip-path: inset(100%);
  clip-path: inset(100%);
  -webkit-clip: rect(1px,1px,1px,1px);
  clip: rect(1px,1px,1px,1px);
  height: 1px;
  overflow: hidden;
  position: absolute;
  width: 1px;
}

.c0 {
  padding: 0.5rem 0.25rem;
  background-color: #FFFFFF;
  display: inline-block;
  font-family: ReithSans,Helvetica,Arial,sans-serif;
  font-size: 0.75rem;
  line-height: 1rem;
  color: #222222;
}

.c1 {
  vertical-align: middle;
  margin: 0 0.25rem;
  fill: #222222;
}

.c2 {
  vertical-align: middle;
  margin: 0 0.25rem;
}

<div
  className="c0"
>
  <svg
    aria-hidden="true"
    className="c1"
    focusable="false"
    height="12px"
    viewBox="0 0 32 32"
    width="12px"
  >
    <polygon
      points="3,32 29,16 3,0"
    />
  </svg>
  <time
    className="c2"
    dateTime="PT2M15S"
  >
    <span
      className="c3"
    >
      Video 2 minutes 15 seconds
    </span>
    <span
      aria-hidden="true"
    >
      2:15
    </span>
  </time>
</div>
`;

exports[`MediaIndicator should render correctly without duration details 1`] = `
.c0 {
  padding: 0.5rem 0.25rem;
  background-color: #FFFFFF;
  display: inline-block;
  font-family: ReithSans,Helvetica,Arial,sans-serif;
  font-size: 0.75rem;
  line-height: 1rem;
  color: #222222;
}

.c1 {
  vertical-align: middle;
  margin: 0 0.25rem;
  fill: #222222;
}

<div
  className="c0"
>
  <svg
    aria-hidden="true"
    className="c1"
    focusable="false"
    height="12px"
    viewBox="0 0 32 32"
    width="12px"
  >
    <polygon
      points="3,32 29,16 3,0"
    />
  </svg>
</div>
`;
